R/tar-stars.R

```diff
@@ -0,0 +1,283 @@
+    options <- options %||% character(0)
+
+    # get drivers available for writing (depends on user's GDAL build)
+    drv <- sf::st_drivers(what = "raster")
+    drv <- drv[drv$write, ]
+
+    driver <- rlang::arg_match0(driver, drv$name)
+
+    .read_stars <- eval(substitute(
+        function(path) {
+
+        ## TODO: it appears envvar custom resources do not work in read function?
+        READ_FUN <- stars::read_stars
+        #  mdim <- as.logical(Sys.getenv("GEOTARGETS_GDAL_RASTER_MDIM", unset = FALSE))
+        if (mdim) {
+            READ_FUN <- stars::read_mdim
+        }
+
+        # ncdf <- as.logical(Sys.getenv("GEOTARGETS_USE_NCMETA", unset = FALSE))
+        if (ncdf && requireNamespace("ncmeta")) {
+            READ_FUN <- stars::read_ncdf
+        }
+
+        # proxy <- as.logical(Sys.getenv("GEOTARGETS_PROXY", unset = FALSE))
+        READ_FUN(path, proxy = proxy)
+
+    }, list(ncdf = ncdf, mdim = mdim, proxy = proxy)))
+
+    # TODO: should multidimensional array use the same `options` as 2D?
+    .write_stars <- function(object, path) {
+
+        WRITE_FUN <- stars::write_stars
+
+        mdim <- as.logical(Sys.getenv("GEOTARGETS_GDAL_RASTER_MDIM",
+                                      unset = FALSE))
+        if (mdim) {
+            WRITE_FUN <- stars::write_mdim
+        }
+
+        dr <- Sys.getenv("GEOTARGETS_GDAL_RASTER_DRIVER")
+
+        # stars requires character(0), not "", for no options set
+        co <- Sys.getenv("GEOTARGETS_GDAL_RASTER_CREATION_OPTIONS")
+        co2 <- strsplit(co, ";")[[1]]
+
+        WRITE_FUN(
+            object,
+            path,
+            overwrite = TRUE,
+            driver = dr,
+            options = co
+        )
+    }
+
+    targets::tar_target_raw(
+        name = name,
+        command = command,
+        pattern = pattern,
+        packages = packages,
+        library = library,
+        format = targets::tar_format(
+            read = .read_stars,
+            write = .write_stars,
+            marshal = function(object) object, # Not currently used
+            unmarshal = function(object) object
+        ),
+        repository = repository,
+        iteration = iteration,
+        error = error,
+        memory = memory,
+        garbage_collection = garbage_collection,
+        deployment = deployment,
+        priority = priority,
+        resources = targets::tar_resources(
+            custom_format = targets::tar_resources_custom_format(
+                #these envvars are used in read and write functions of format
+                envvars = c("GEOTARGETS_GDAL_RASTER_DRIVER" = driver,
+                            "GEOTARGETS_GDAL_RASTER_CREATION_OPTIONS" =
+                                paste0(options, collapse = ";"),
+                            "GEOTARGETS_GDAL_RASTER_MDIM" = mdim,
+                            "GEOTARGETS_PROXY" = proxy,
+                            "GEOTARGETS_USE_NCMETA" = ncdf)
+            )
+        ),
```

Comment:
Use user-specified `resources` as in #67 
```suggestion
        resources = utils::modifyList(targets::tar_resources(
            custom_format = targets::tar_resources_custom_format(
                #these envvars are used in read and write functions of format
                envvars = c("GEOTARGETS_GDAL_RASTER_DRIVER" = driver,
                            "GEOTARGETS_GDAL_RASTER_CREATION_OPTIONS" =
                                paste0(options, collapse = ";"),
                            "GEOTARGETS_GDAL_RASTER_MDIM" = mdim,
                            "GEOTARGETS_PROXY" = proxy,
                            "GEOTARGETS_USE_NCMETA" = ncdf)
            )
        ), resources)
```